### PR TITLE
Create framework for tests that analyze HTTP communication

### PR DESCRIPTION
**Pull request by wkkasztan**
In such tests, driver communicates with alternator, while a proxy is listening for and possibly modifying messages in between.

### CODE REVIEW

**Review by lukaszg22 — APPROVED** [review]

**Review by m-szymon — CHANGES_REQUESTED** [review]
In general I would expect a better example of a real world test.

File: tests/empty_test.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+use hyper::client::conn::http1::SendRequest;
+use hyper::{Request, Response};
+
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+mod common;
+
+#[tokio::test]
+async fn empty_test() {
```
[the remark this answers is not in the document]
Comment (wkkasztan) [reply tests/empty_test.rs]: Each test from the "HTTP traffic - parsing" family is based on the `empty_test`.
It is there to make others easier to understand, by showing a simple working example of a test that uses proxy between driver and alternator.
Though, it also does test wheter the proxy system works (as all of the other tests).

I can remove it if it's not needed, or change it's name to `proxy_test`, though i think `empty_test` or `example_test` is a better name, as it suggests simplicity.
Or we can forward this problem to the next pull request, when there are more tests to see.

File: tests/common.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+pub async fn acquire_dynamodb_response(
+    parts: http::response::Parts,
+    body: Bytes,
+) -> Response<Full<Bytes>> {
+    let body = Full::new(body);
+    Response::from_parts(parts, body)
+}
+
+pub async fn make_calls(client: dynamodb::Client) {
```
[the remark this answers is not in the document]
Comment (m-szymon) [reply tests/common.rs]: I'm in favor of tests where the core test is implemented in the main function. We can have helper functions that perform setup/cleanup, which is not the subject of the test.
If we want to test CreateTable, we don't call the helper function that does this, but directly in the main function.

File: tests/empty_test.rs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    let config = dynamodb::Config::builder()
+        .endpoint_url("http://localhost:7999")
+        .credentials_provider(dynamodb::config::Credentials::for_tests_with_session_token())
+        .region(dynamodb::config::Region::new("eu-central-1"))
+        .behavior_version(dynamodb::config::BehaviorVersion::latest())
+        .build();
+
+    let client = dynamodb::Client::from_conf(config);
+
+    let proxy = common::start_server("localhost:7999", "localhost:8000").await;
```
[the remark this answers is not in the document]
Comment (m-szymon) [reply tests/empty_test.rs]: I meant 7999 - it is not related to docker.

**Review by m-szymon — CHANGES_REQUESTED** [review]
I am very picky about this PR, because I believe tests should be really simple and easy to write and understand, to encourage writing lot of them with many corner cases and to be helpful for anyone who touches the code next.
Currently I don't see it like that.

I still insist that the "triggers" are too much separated from "checks". It is not that actual test happens in proxy handler. The trigger is a command that we call and it would be nice to see its expected results after it. More complex tests that you showed me only confirm my thoughts.

Maybe consider using `TaskSet` to more explicitly manage lifetime of task and handling errors and asserts. 

I suggest renaming `empty_test.rs` to `proxy_test.rs` and having several very simple tests showing (and proving) how you can work with proxy. Ideas:
- current basic test - simplify it even more - get empty table list from alternator, but call it explicitly.
- basic mock - below I suggested how to use it as a mock - again show it with just empty table list or even better http error code and how to expect it on client side.
- test with pre-creating table - how to handle resources. Again actual test covers single command.
- your `correct_line` test - it shows asserts directly in proxy handler - how is it handed.
- simple test showing pattern that you have in `body_compression` - shared state updated by proxy handler across several commands and asserted at the end in main task. E.g. count the number of requests.

Consider also timeouts. I guess now it relies on client timeout - something we may actually want to test. Are there risks of hangs?

**Review by m-szymon — CHANGES_REQUESTED** [review]
In general looks good.
Several comments below - but they me be applicable in several places.

I think you should split commits or even PRs.
Basic proxy+http_test+correct_line (as a simple example) could be merged.
The specific tests need some better defining first in test plan and can be added in individual commits.

File: tests/http_content/common/http_test.rs [hunk 4]
```diff
@@ -0,0 +1,236 @@
+//! `HttpTest` provides a proxy-based environment for intercepting and asserting
+//! HTTP traffic between a driver and alternator.
```
Comment (m-szymon) [inline tests/http_content/common/http_test.rs]: Maybe "HttpTester" is more appropriate?
As far as I remember previously it was a test class that you had to override, right?
Now it is rather a tool that you use in a test (which is very good).

**Review by m-szymon — CHANGES_REQUESTED** [review]
It looks very good. This is the path to follow. Good job.

One small issue is to reconsider cleanup workflow.
In this commit we can have some explicit handling.
Then depending on what scenarios we implement next, we could have a commit adding some helpers.
But tracking and handling of resources to cleanup is not responsibility of tools we create here.

Other request is to add some more description: to a test and in the commit/PR message.

**Review by m-szymon — APPROVED** [review]
As I said before conceptually cleanup shouldn't be to the task for HttpTester. 
But I see implementation issues.
Cleanest solution I see would be something like 'test-context' crate. Such context could track resources to be cleaned and encapsulate common setup/teardown actions.
But it is not critical - something to consider when the number of tests grows.

File: tests/http_content/common/http_test.rs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+//!
+//! 3. Inside the test function, construct a new driver client with
+//!    endpoint HttpTestContext::get_proxy_address().
+//!
+//! 4. Whenever you create a resource that requires cleanup,
+//!    register it with HttpTestContext::register_resource()
+//!
+//! 5. You can override the initial on_request with HttpTestContext::set_on_request().
+#![allow(dead_code)]
+
```
Comment (m-szymon) [inline tests/http_content/common/http_test.rs:25]: Is `#![allow(dead_code)]` the right approach here? It's a blanket suppression for the whole file — wouldn't it risk hiding genuinely unused items too? I wonder if the root cause is the `#[path = ...]` include pattern, which means this module gets compiled separately into each test binary, and `pub` items not used in that particular binary trigger dead_code warnings. If that's the case, maybe it's worth a short comment explaining why this allow is here, rather than leaving it implicit?

**Review by m-szymon — APPROVED** [review]
Don't merge master, but rebase on master.
When you fix that I will merge.

File: tests/http_content/common/http_test.rs (hunk 5 above)
Comment (wkkasztan) [reply tests/http_content/common/http_test.rs:25]: The issue is with `HttpTestContext::set_on_request`, which isn't used for correct_line test, but is crucial for the next tests. I should have marked this specific method as dead code instead of whole file
